src/dbnode/storage/limits/query_limits.go

```diff
@@ -0,0 +1,243 @@
+	Stop()
+}
+
+// LookbackLimit provides an interface for a specific query limit.
+type LookbackLimit interface {
+	Inc(new int) error
+	Exceeded() error
+	Start()
+	Stop()
+
+	// For testing purposes.
+	current() int64
+	reset()
+}
+
+// LookbackLimitOptions holds options for a lookback limit to be enforced.
+type LookbackLimitOptions struct {
+	// Limit past which errors will be returned.
+	Limit int64
+	// Lookback is the period over which the limit is enforced.
+	Lookback time.Duration
+}
+
+// NewQueryLimits returns a new query limits manager.
+func NewQueryLimits(
+	instrumentOpts instrument.Options,
+	docsLimitOpts LookbackLimitOptions,
+	bytesReadLimitOpts LookbackLimitOptions,
+) QueryLimits {
+	scope := instrumentOpts.
+		MetricsScope().
+		SubScope("query-limits")
+	docsLimit := NewLookbackLimit("docs-matched", scope, docsLimitOpts)
+	bytesReadLimit := NewLookbackLimit("bytes-read", scope, bytesReadLimitOpts)
+	return &queryLimits{
+		scope:          scope,
+		docsLimit:      docsLimit,
+		bytesReadLimit: bytesReadLimit,
+	}
+}
+
+// NewLookbackLimit returns a new lookback limit.
+func NewLookbackLimit(name string, scope tally.Scope, opts LookbackLimitOptions) LookbackLimit {
+	return &lookbackLimit{
+		name:    name,
+		options: opts,
+		metrics: lookbackLimitMetrics{
+			recent:     scope.Gauge(fmt.Sprintf("recent-%s", name)),
+			recentPeak: scope.Gauge(fmt.Sprintf("recent-peak-%s", name)),
+			total:      scope.Counter(fmt.Sprintf("total-%s", name)),
+			exceeded:   scope.Tagged(map[string]string{"limit": name}).Counter("limit-exceeded"),
+		},
+		recent: atomic.NewInt64(0),
+		stopCh: make(chan struct{}),
+	}
+}
+
+func (q *queryLimits) DocsLimit() LookbackLimit {
+	return q.docsLimit
+}
+
+func (q *queryLimits) BytesReadLimit() LookbackLimit {
+	return q.bytesReadLimit
+}
+
+func (q *queryLimits) Start() {
+	q.docsLimit.Start()
+	q.bytesReadLimit.Start()
+}
+
+func (q *queryLimits) Stop() {
+	q.docsLimit.Stop()
+	q.bytesReadLimit.Stop()
+}
+
+func (q *queryLimits) AnyExceeded() error {
+	if err := q.docsLimit.Exceeded(); err != nil {
+		return err
+	}
+	if err := q.bytesReadLimit.Exceeded(); err != nil {
+		return err
+	}
+	return nil
+}
+
+// Inc increments the current value and returns an error if above the limit.
+func (q *lookbackLimit) Inc(new int) error {
+	if q == nil {
+		return nil
+	}
+	if new < 0 {
+		return fmt.Errorf("invalid negative query limit inc %d", new)
+	}
+	if new == 0 {
+		return q.Exceeded()
+	}
+
+	// Add the new stats to the global state.
+	newI64 := int64(new)
+	recent := q.recent.Add(newI64)
+
+	// Update metrics.
+	q.metrics.recent.Update(float64(recent))
+	q.metrics.total.Inc(newI64)
+
+	// Enforce limit (if specified).
+	return q.exceeded(recent)
+}
+
+func (q *lookbackLimit) Exceeded() error {
+	return q.exceeded(q.recent.Load())
+}
+
+func (q *lookbackLimit) exceeded(recent int64) error {
+	if q.options.Limit > 0 && recent > q.options.Limit {
+		q.metrics.exceeded.Inc(1)
+		return fmt.Errorf(
```

Comment:
Just a sanity check that if e.g. the first series out of 100,000 trips this we won't end up with 100,000 errors added to the logs?

Reply:
What if this trips: https://github.com/m3db/m3/pull/2627/files#diff-667e8a7e7c5485bf7d69c2f73cbd4887R311

Reply:
This is probably a more general question about the downstream of this, just to ensure it's using the correct `mutliError` return format

Reply:
Basically these `req` instances get passed back up as `SegmentReader`s:
```
return xio.BlockReader{
		SegmentReader: req,
		Start:         req.start,
		BlockSize:     req.blockSize,
	}
```

And then if they ever are used w/ an error assign, they fail. E.g.
```
func (req *retrieveRequest) Read(b []byte) (int, error) {
	req.resultWg.Wait()
	if req.err != nil {
		return 0, req.err
	}
	return req.reader.Read(b)
}

func (req *retrieveRequest) Segment() (ts.Segment, error) {
	req.resultWg.Wait()
	if req.err != nil {
		return ts.Segment{}, req.err
	}
	return req.reader.Segment()
}
```
And upstream when iterating these readers it always just bails after first one. So this all seems to do what we expect in that it fails on a per query basis. I can also verify though in the logs when retesting this on rc-05

Reply:
Good stuff, just wanting to avoid a ton of spam here :+1: